plugins/source/pagerduty/codegen/recipes/base.go

```diff
@@ -0,0 +1,358 @@
+	if err != nil {
+		fmt.Printf("failed to format source: %s: %v\n", filePath, err)
+	} else {
+		content = formattedContent
+	}
+	if err := os.WriteFile(filePath, content, 0644); err != nil {
+		return fmt.Errorf("failed to write file %s: %w", filePath, err)
+	}
+
+	return nil
+}
+
+// SetParentChildRelationships calculates and sets the parent and children fields on resources.
+func SetParentChildRelationships(resources []*Resource) error {
+	m := map[string]*Resource{}
+	for _, r := range resources {
+		m[r.SubService] = r
+	}
+	csr := caser.New()
+	for _, r := range resources {
+		for _, ch := range r.Relations {
+			name := csr.ToSnake(strings.TrimSuffix(ch, "()"))
+			v, ok := m[name]
+			if !ok {
+				return errors.New("child not found for " + r.SubService + " : " + name)
+			}
+			r.Children = append(r.Children, v)
+			v.Parent = r
+		}
+	}
+	return nil
+}
+
+func GenerateAllTablesList(resources []*Resource) error {
+	tpl, err := template.New("all_tables.go.tpl").Funcs(template.FuncMap{
+		"ToCamel": strcase.ToCamel,
+		"ToLower": strings.ToLower,
+	}).ParseFS(templatesFS, "templates/all_tables.go.tpl")
+	if err != nil {
+		return fmt.Errorf("failed to parse templates: %w", err)
+	}
+
+	var buff bytes.Buffer
+	if err := tpl.Execute(&buff, filterRootResources(resources)); err != nil {
+		return fmt.Errorf("failed to execute template: %w", err)
+	}
+
+	_, runtime_filename, _, ok := runtime.Caller(0) // Make codegen work independently of `pwd`.
+	if !ok {
+		return fmt.Errorf("failed to get caller information")
+	}
+
+	filePath := path.Join(path.Dir(runtime_filename), "../../", "resources/plugin", "all_tables.go")
+	content := buff.Bytes()
+	formattedContent, err := format.Source(buff.Bytes())
+	if err != nil {
+		fmt.Printf("failed to format source: %s: %v\n", filePath, err)
+	} else {
+		content = formattedContent
+	}
+	if err := os.WriteFile(filePath, content, 0644); err != nil {
+		return fmt.Errorf("failed to write file %s: %w", filePath, err)
+	}
+
+	return nil
+}
+
+// filterRootResources returns only root resources from the input list
+func filterRootResources(resources []*Resource) []*Resource {
+	result := make([]*Resource, 0)
+
+	for _, r := range resources {
+		if r.Parent == nil {
+			result = append(result, r)
+		}
+	}
+
+	return result
+}
+
+// Golang hates files with `_windows.go` extension...
+// This function replaces such extension with `_window.go`.
+func MaybeReplaceUnderscoreWindowsSuffix(str string) string {
+	return strings.ReplaceAll(str, "_windows.go", "_window.go")
+}
+
+// The default type-transfomer, but replace `created_at`, `deleted_at` fields with `TypeTimestamp` instead
+// of `TypeString`.
+func pagerDutyTypeTransformer(field reflect.StructField) (schema.ValueType, error) {
+	if !isStringField(field.Type) {
+		return codegen.DefaultTypeTransformer(field)
+	}
+
+	timestampFieldNames := []string{
+		"CreateAt", "CreatedAt", "DeletedAt", "LastStatusChangeAt", "StartTime", "EndTime", "LastIncidentTimestamp",
```

Comment:
can we actually check for underlying type instead?

Reply:
because if any new field appears with a different name we'll have a problem

Reply:
Not easily - the underlying types are `string`...
I think it's alright for now.